### DISCUSSION TITLE
asyncio.start_server, number of connections and then lock-up

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#15762] [Using MicroPython] [answered]
I am running a simple asyncio server listening for several clients.  One of the clients (socket and not asyncio-based) sends a brief message and then closes.  If I hit the server with multiple messages, say 10-15 seconds apart, the server appears to lock-up after 6, 7, 8 or 9 attempts.
```
awaiting client connection ...
Received 'gate alarm' from ('192.168.8.60', 64294)
Received 'gate alarm' from ('192.168.8.60', 64295)
Received 'gate alarm' from ('192.168.8.60', 64296)
Received 'gate alarm' from ('192.168.8.60', 64297)
Received 'gate alarm' from ('192.168.8.60', 64298)
Received 'gate alarm' from ('192.168.8.60', 64299)
Received 'gate alarm' from ('192.168.8.60', 64300)
Received 'gate alarm' from ('192.168.8.60', 64301)
Received 'gate alarm' from ('192.168.8.60', 64302)
```
Sometimes after 5 minutes I can start sending messages again and sometimes after waiting about 20 minutes I get:
`[Error 113] ECONNABORTED` 
and then I can start sending messages again.

Should I be configuring something to drop the connection after use?  

Thanks for any keywords I can read-up on.

Is this where I need to look:
https://stackoverflow.com/questions/53950344/how-to-implement-timeout-in-asyncio-server

Or better still?
https://superfastpython.com/asyncio-timeout-best-practices/

**Comment by DavesCodeMusings** [comment]
Do you have code to close the connection when you're done? Something like below.

```py
async def on_connect(self, reader, writer):
    # Read request and send reply.

    # Then close connection.
    await writer.drain()
    writer.close()
    await writer.wait_closed()
    reader.close()
    await reader.wait_closed()
```

**Reply by davefes** [reply]
Ah, I was only closing the connection on `except`
```py
async def handle_echo(reader, writer):
    addr = writer.get_extra_info('peername')

    try:
        while (data := await reader.read(100)):  # walrus operator, needs the parentheses
            message = data.decode()
            print(f'Received {message!r} from {addr!r}')

            if message == 'ping' and config.watchdog is True:
                print('got a ping')

                wdt_feed()
                wdt.feed()

            elif message == 'you have got a gate IR alarm':
                print('you got a gate alarm')

            print(f'Send: {message!r}')
            writer.write(data)
            await writer.drain()

    except Exception as err:
        print(str(err))
        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write(str(err) + '\n')
        except OSError:
            pass

        print('closing the connection...')
        writer.close()
        await writer.wait_closed()
        print('closed')

        print('doing a machine.reset()')
        reset()
```
I'll re-think this.  Thanks.

**Reply by davefes** [reply]
My thinking was that for `ping` I didn't want to close the connection.  They happen every 20 seconds, although I'd be happy making that 60 seconds or more.  Should I close after every message?

The `ping` client is asyncio-based, whereas the gate alarm is a  "normal" sync-based socket connection.  Are you allowed to mix them on the one server?

I noticed that the ping `peername` did not change, whereas the gate alarm `peername` was changing each time it was sent.

**Reply by DavesCodeMusings** [reply, marked as answer]
My experience with asyncio thus far has been with an HTTP server, that closes the connection after each request is serviced, and an FTP server that keeps a control open until the user logs out. So, you can do either. But, you can only have so many open connections. I think the default is five, but I could be wrong.

If the client is making a connection, sending something, then closing the connection, you should be closing on the server side as well. Otherwise, you're going to end up with a bunch of connections tied up in a FINWAIT state. If you've got a tool like netstat on your client side, you can see it. Eventually, the client will time out and drop these. That's probably why it starts working again if you wait a few minutes between attempts.

**Reply by davefes** [reply]
I close after every transaction and that sorts it.

Thanks again for taking the time to reply ... I have been battling this one for quite awhile!

**Reply by DavesCodeMusings** [reply]
You could have the ping client do open, send, close as well. I think unless you're trying to maintain a user session or some kind of state information, you may as well close when you're done. Best of luck!

**Reply by davefes** [reply]
I just realised that was the difference, I will look at closing the ping client after every send.
